src/classes/AiActions/AiHandler.js

```diff
@@ -0,0 +1,98 @@
+    let boolSide = store.getters.getCoinMsg
+    let canGroup = this.move.findGroup(event.stack, hand.bestGCard)
+    this.antiVirus = new AntiVirus(hand, boolSide, this.move, event)
+    this.firewall = new Firewall(hand, boolSide, this.move, event)
+    this.generator = new Generator(hand, boolSide, this.move, event)
+    this.overclock = new OverClock(hand, boolSide, this.move, event)
+    this.batteryBackup = new BatteryBackup(hand, boolSide, this.move, event)
+    this.group = new Group(hand, boolSide, this.move, event, canGroup)
+    this.virus = new Virus(hand, boolSide, this.move, event, this.opponentVirus)
+    this.hack = new Hack(hand, boolSide, this.move, event)
+    this.powerOutage = new PowerOutage(hand, boolSide, this.move, event, this.opponentPO)
+    this.instruction = new Instruction(hand, boolSide, this.move, event)
+    this.repeat = new Repeat(hand, boolSide, this.move, event)
+    this.repeatX = new RepeatX(hand, boolSide, this.move, event)
+    this.variable = new Variable(hand, boolSide, this.move, event)
+    this.discard = new Discard(hand, boolSide, this.move, event)
+    this.stackToPlay = undefined
+    this.cardToPlay = undefined
+    this.opponentToAttack = undefined
+    this.moveType = undefined
+  }
+
+  /**
+   * This creates the order of operations for the ai steps and executes them until a step has been chosen.
+   * @param type The type of Ai.
+   */
+  setAi (type) {
+    let moveList = []
+    if (type === 'gambler') {
+      moveList = [this.hack, this.virus, this.powerOutage, this.firewall, this.generator, this.antiVirus, this.overclock,
+        this.batteryBackup, this.variable, this.repeat, this.instruction, this.repeatX, this.group, this.discard]
+    } else if (type === 'hacker') {
+      moveList = [this.hack, this.virus, this.powerOutage, this.firewall, this.generator, this.antiVirus, this.overclock, this.batteryBackup, this.variable, this.repeat,
+        this.instruction, this.repeatX, this.group, this.discard]
+    } else if (type === 'sprinter') {
+      moveList = [this.variable, this.overclock, this.repeatX, this.repeat, this.virus, this.firewall, this.generator, this.antiVirus, this.powerOutage, this.batteryBackup,
+        this.instruction, this.hack, this.group, this.instruction, this.discard]
+    } else if (type === 'protector') {
+      moveList = [this.group, this.firewall, this.generator, this.antiVirus, this.overclock, this.virus, this.powerOutage, this.batteryBackup, this.variable,
+        this.repeat, this.instruction, this.repeatX, this.hack, this.instruction, this.discard]
+    }
+    for (let turn in moveList) {
+      if (moveList[turn].execute()) {
+        this.stackToPlay = moveList[turn].getStack()
+        this.cardToPlay = moveList[turn].getCard()
+        this.opponentToAttack = moveList[turn].getOpponent()
+        this.moveType = moveList[turn].getMove()
+        break
+      }
+    }
+  }
+  getMove () {
+    return this.moveType
```

Comment:
I like this idea of `move type`. I think we can take this a step further in the refactoring. The idea is to put all of the `protection` moves into a single array. The AIs would then always first protect themselves (i.e. go  through this move array) before doing any other moves. 

I think the advantage here is that the order of protection moves will always be the same anyways (or can  easily be changed in the future for all AIs), so we abstract this commonality. This will also help to better highlight the differences in move priorities for different AIs in the code.

What do you  think? Let me know if you need clarification about this.

Reply:
This is a good idea. Just to clarify an example of this would be a safetyMoves = [this.firewall, this.generator, this.antiVirus] and then each individual ai's moveList would look something like this: 
moveList = [safetyMoves, protectionMoves, attackMoves, playMoves] and then collapse the 2d array into a 1d array?

Reply:
Reading through this again I think I misunderstood. Do you only want the protection/safety card into their own array or other types as well?

Reply:
I was just thinking about refactoring out the safety and protection moves into an array, although you could create an array for each. I think we need to leave the attack and play moves as a single array, as this is where the AIs would vary.

The idea is to have something like this:
```foreach (action : protection/safety)
   if(action) break
foreach(action : attack/play)
   if(action) break
discard

Reply:
I think I implemented this correctly in the latest commit. Instead of running a separate for loop for the protection/safety cards I added them to the beginning of the moveList array to avoid repeated code. Let me know if this fixes it.